### DISCUSSION TITLE
configuring in an initializer?

### DISCUSSION THREAD

**Discussion by jrochkind** [bensheldon/good_job#626] [Q&A] [answered]
The [README docs](https://github.com/bensheldon/good_job#configuration-options):

> GoodJob configuration can be placed within Rails config directory for all environments (config/application.rb), within a particular environment (e.g. config/environments/development.rb), or within an initializer (e.g. config/initializers/good_job.rb).

One example of doing it in an initializer would look like:

    GoodJob.preserve_job_records = true

But if I want to also set `cleanup_discarded_jobs` (which frankly seems to me like it should be the default recommended configuration, both `preserve_job_records` and `cleanup_discarded_jobs`; it's clear to me that is what I want, but it was a bit confusing to find)....

`GoodJob.cleanup_discarded_jobs = true` doesn't work.  Is there any good way to set `cleanup_discarded_jobs` in an initializer?

Some configuration options seem to be available off `GoodJob`, but others not?  I would love to put all my cross-environment GoodJob config in an initializer.

**Comment by bensheldon** [comment, marked as answer]
It's safe to use all of the `config.good_job....` settings inside of an initializer: https://github.com/bensheldon/good_job/issues/380#issuecomment-1103152775

I'd recommend entirely using the config settings, and maybe the Readme should reflect that.

I think the only `GoodJob...` global setting that _doesn't_ have an equivalent in `config.good_job...` is `GoodJob.active_record_parent_class`.

I'd love feedback of how to make it clearer in the Readme. Also, the defaults will become better in [GoodJob 3.0](https://github.com/bensheldon/good_job/issues/507).

**Reply by jrochkind** [reply]
Thanks, can you give an example of how to use the `config.good_job` settings in an initializer?

You can't just refer to `config` in an initializer, it doesn't exist. 

Are you suggesting, uh... `Rails.application.config.good_job.cleanup_discarded_jobs = false`?  That seems to maybe work, at least it doesn't error.... do you think that's safe?

I'm not sure most people would come up with that one. The README still includes some examples of configuring `GoodJob.something` directly... Search REAMDE for `You’ll generally want to configure these in config/initializers/good_job.rb, like so`. Are you saying that example is bad advice, and shoudl be advising... `Rails.application.config.good_job`? 

Good to hear defaults will be better in 3.0!

**Reply by bensheldon** [reply]
You can use this pattern in an initializer:

```ruby
Rails.application.configure do
  config.good_job.cleanup_discarded_jobs = false
end
```

This is perfectly safe for GoodJob, but that required intentionality on my part.  It's a little problematic because it's not universal; see https://github.com/rails/rails/issues/36650

GoodJob's development harness uses this pattern: https://github.com/bensheldon/good_job/blob/05300e16c5fb172f4b94f3ed9a46b3fa735fc3d6/spec/test_app/config/initializers/good_job.rb

**Reply by bensheldon** [reply]
Oh, and thank you, incredibly much, for the feedback about the Readme. That needs to be improved because I'm sure plenty of people are confused if you're asking the question.

**Reply by jrochkind** [reply]
Hey thanks for guidance! taht link to the test app initializer is useful. 

I could PR some improvements to README... but I'm still confused. 

The test app [initializer you link to](https://github.com/bensheldon/good_job/blob/05300e16c5fb172f4b94f3ed9a46b3fa735fc3d6/spec/test_app/config/initializers/good_job.rb#L18-L20) for instance  still has in it:

```ruby
  GoodJob.retry_on_unhandled_error = false
  GoodJob.preserve_job_records = true
  GoodJob.on_thread_error = -> (error) { Rails.logger.warn("#{error}\n#{error.backtrace}") }
```

Should these all preferably be included in the `Rails.application.configure`  block below as eg `config.good_job.preserve_job_records = true` instead?  

Or are there some configuration that should still be done on `GoodJob`?  

Maybe not, I see it _is_ given with example config.good_job.preserve_job_records` in README at [Configuration options](https://github.com/bensheldon/good_job/blob/main/README.md#configuration-options). 

OK, should I PR to README changing _all_ references to any config to use the `Rails.application.config` format instead?  (And update test app initializer too, to use new preferred form?)

In my branch of my actual app where I'm testing out good_job, I'm currently having trouble getting `preserve_job_records` to work at all, no matter what form of config -- not sure what's going on there, I had it working a couple days ago!

**Reply by bensheldon** [reply]
I'm hoping to release GoodJob 3.0 this weekend, so let's hold off on updating the Readme until then.

The only configuration that _must_ be done on GoodJob is `GoodJob.active_record_parent_class`. All other configuration can be set through `Rails.application.config`.

> In my branch of my actual app where I'm testing out good_job, I'm currently having trouble getting `preserve_job_records` to work at all

😬 Would you be able to share the project with me? I'm happy to dig in. I feel somewhat confident that things should work because https://goodjob-demo.herokuapp.com/good_job/jobs is continuously deployed with the most recent version of GoodJob... but definitely something I could have missed or a weird interaction. btw, looking at the code, currently `Rails.application.config` takes precedence over config on `GoodJob`:

https://github.com/bensheldon/good_job/blob/caa7378480c24124dd32675d0fe4474e20f383fa/lib/good_job/engine.rb#L28-L35

**Reply by jrochkind** [reply]
Why wouldn't we want to update the README before the 3.0 release, to get in there first? If the `Rails.application.config` is the preferred way to do it, for consistency accross all config? But maybe that's not actually the preferred way? 

I was confused with some config elements being set one way and some another, and it not being documented which should be set which way -- but I am just one person, I may not be typical! Either way, okay, your call, if you don't want to make the README consistent in that way?

Good to know good_job 3.0 is coming shortly -- I will perhaps pause our attempt to integrate good_job in our app until 3.0 comes out, and start with that one! No reason to get everything in order for 2.0 now and then just have to upgrade and change things! 

good_job is looking pretty nice! Except I'm still a bit baffled at the lack of First-in-first-out mentioned in https://github.com/bensheldon/good_job/discussions/624 .  I'm not sure if I'm comfortable switching over to a queue that isn't first-in-first-out... although I guess if nobody else has wanted that or noticed maybe I'm wrong that it matters that much, but it seems disturbing to me.  Also to fully migrate we'd need to, I guess, assign priorities to everything that we were currently use queue-order-of-precedence to handle, since good_job doesn't do that, as mentioned in that discussion.

**Reply by bensheldon** [reply]
Sorry, didn't mean to throw cold water on your work right now. 3.0 will not have any major changes, just removing deprecations and swapping some of the behaviors (preserve jobs, and auto-delete them will become the default).

I mostly don't want you to go to the trouble of updating the readme, and then those sections being almost immediately deleted because they won't be necessary. Right now a not-insignificant part of the readme is focused on configurating towards what in 3.0 will be the default.

But also, if I hit my deadline, I assume that waiting a week won't block you.

I will reply more deeply on the FIFO issue. I really hasn't come up, but I can explain more over there.

**Reply by jrochkind** [reply]
Ah, I get it. 

So the current README doesn't have the 3.0 chagnes yet... do you have a branch that does that I can PR against, or you just haven't updated the README for 3.0 yet?

**Reply by bensheldon** [reply]
My doctrine has been to only use major releases for deprecations and functional changes (e.g. changing preserve-jobs default from false to true) not to introduce new features. So everything that will exist in 3.0 currently exists now, it just may be disabled by default.

The functional changes for 3.0 are in https://github.com/bensheldon/good_job/pull/545, which is simply changing the default values.  I'll also make 1 more PR which removes deprecations. And that's it. (unless there are any unconsidered surprises from that).

**Reply by bensheldon** [reply]
And I'll add that the 1.0 -> 2.0 upgrade cycle had a change to the locking strategy that required an orchestrated production migration. 2.0 -> 3.0 doesn't have anything like that.
